source/ship/ShipAICache.cpp

```diff
@@ -0,0 +1,115 @@
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE. See the GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License along with
+this program. If not, see <https://www.gnu.org/licenses/>.
+*/
+
+#include "ShipAICache.h"
+
+#include "../pi.h"
+#include "../Ship.h"
+
+#include <algorithm>
+#include <cmath>
+
+using namespace std;
+
+
+
+// Constructor.
+ShipAICache::ShipAICache(const Ship &ship)
+{
+	this->ship = &ship;
+	UpdateWeaponCache();
+}
+
+
+
+void ShipAICache::UpdateWeaponCache()
+{
+	bool hasAmmo = false;
+	bool isArmed = false;
+	double totalDPS = 0.;
+	double splashDPS = 0.;
+	double artilleryDPS = 0.;
+
+	shortestRange = 4000.;
+	shortestArtillery = 4000.;
+	minSafeDistance = 0.;
+
+	for(const Hardpoint &hardpoint : ship->Weapons())
+	{
+		const Outfit *weapon = hardpoint.GetOutfit();
+		if(weapon && !hardpoint.IsAntiMissile())
+		{
+			isArmed = true;
+			bool lackingAmmo = (weapon->Ammo() && weapon->AmmoUsage() && !ship->OutfitCount(weapon->Ammo()));
+			// Weapons without ammo might as well not exsit, so don't even consider them
+			if(lackingAmmo)
+				continue;
+			hasAmmo = true;
+
+			// Calculate the damage per second, ignoring any special effects. (could be improved to account for those)
+			double DPS = (weapon->ShieldDamage() + weapon->HullDamage()
+				+ (weapon->RelativeShieldDamage() * ship->Attributes().Get("shields"))
+				+ (weapon->RelativeHullDamage() * ship->Attributes().Get("hull")))
+				/ weapon->Reload();
+			totalDPS += DPS;
+
+			// Exploding weaponry that can damage this ship requires special consideration.
+			if(weapon->SafeRange())
+			{
+				minSafeDistance = max(weapon->SafeRange(), minSafeDistance);
+				splashDPS += DPS;
+			}
+
+			// The artillery AI should be applied at 1000 pixels range, or 500 if the weapon is homing.
+			double range = weapon->Range();
+			shortestRange = min(range, shortestRange);
+			if(range >= 1000. || (weapon->Homing() && range >= 500.))
+			{
+				shortestArtillery = min(range, shortestArtillery);
+				artilleryDPS += DPS;
+			}
+		}
+	}
+
+	// Calculate this ship's "turning radius"; that is, the smallest circle it
+	// can make while at full speed.
+	double stepsInHalfTurn = 180. / ship->TurnRate();
+	double circumference = stepsInHalfTurn * ship->MaxVelocity();
+	turningRadius = circumference / PI;
+
+	// If this ship was using the artillery AI to run away and bombard its
+	// target from a distance, have it stop running once it is out of ammo. This
+	// is not realistic, but it's less annoying for the player.
+	if(isArmed && !hasAmmo && !ship->IsYours())
+	{
+		shortestRange = 0.;
+		shortestArtillery = 0.;
```

Comment:
what about the other weapons of the ship?
if a ship runs out of ammo, it could still have long range weapons

Reply:
didnt you say they didnt suicide if they still had long range weapons though?

Reply:
yea, but how is this correlating with the code?
also they still had long range weapons, but they stayed in range for the long range weapon that ran out of ammo, so they were not able to use them

Reply:
oh thats weird they should ignore no-ammo weaps in the range calculations

Reply:
I think I spotted the problem: if the ship has a weapon that uses ammo and is long range and the ship has a weapon that uses ammo and is short range, this code will still consider the long range weapon because there is only one bool for all ammo based weapons

Reply:
hasAmmo will be true if any weapon is true, I'll rename it

Reply:
👍